### PR TITLE
Add speech dataset for wav2vec2 pretraining.

### PR DESCRIPTION
**Pull request by kauterry**
**What does this PR do? Please describe:**
Run this code snippet for testing
```
import torch

from fairseq2.datasets.speech import load_speech_dataset
from fairseq2.gang import setup_default_gang

dataset = load_speech_dataset("librispeech_960h")

gang = setup_default_gang()

data_reader = dataset.create_reader(
    "train",
    gang,
    dtype=torch.float16,
    min_audio_len=32000,
    max_audio_len=250000,
    max_num_elements=1400000,
    normalize_audio=False,
    example_shuffle_window=1000,
    batch_shuffle_window=1000,
    num_accumulate=1,
    num_prefetch=4,
    seed=2,
)

while True:
    try:
        batch = next(data_reader)
        print(batch[0].shape)
    except StopIteration:
        print("End of data reached.")
        break
```
Output:
torch.Size([5, 240320])
torch.Size([5, 242000])
torch.Size([5, 245840])
torch.Size([5, 235040])
torch.Size([5, 237360])
torch.Size([6, 200720])
torch.Size([8, 140800])
torch.Size([5, 234640])
torch.Size([6, 218720])
torch.Size([5, 239840])
torch.Size([8, 128240])
torch.Size([24, 56320])
torch.Size([7, 176561])
torch.Size([5, 233680])
torch.Size([8, 128240])
torch.Size([8, 155920])
torch.Size([5, 233360])

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/datasets/speech.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+
+load_speech_dataset = DelegatingDatasetLoader[SpeechDataset]()
+
+
+# TODO: FIX, INFER
+npc = 10
+
+
+# TODO: Work in progress!
```
Comment (kauterry) [inline src/fairseq2/datasets/speech.py:97]: Not anymore for this right?

File: src/fairseq2/datasets/speech.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        file_mapper = FileMapper(root_data_dir, cached_fd_count=cached_fd_count)
+
+        builder.map(file_mapper, selector="[*].audio")
+
+        # Decode audio.
+        audio_decoder = AudioDecoder(dtype=torch.float32 if normalize_audio else dtype)
+
+        builder.map(audio_decoder, selector="[*].audio.data")
+
+        # TODO(balioglu): Check/adjust sample size
```
Comment (kauterry) [inline src/fairseq2/datasets/speech.py:198]: What dyou mean? This doesn't apply for this dataset right?

File: src/fairseq2/datasets/speech.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        builder.skip(1)
+
+        field_splitter = StrSplitter(names=["audio", "audio_size"])
+
+        builder.map(field_splitter, num_parallel_calls=npc)
+
+        # Cast audio size to integer.
+        builder.map(int, selector="audio_size")
+
+        # TODO: Use `cache()` op.
```
Comment (kauterry) [inline src/fairseq2/datasets/speech.py:285]: Not anymore since we don't load the entire manifest in memory right?

File: src/fairseq2/datasets/speech.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            The name of the dataset.
+        :param manifest_dir:
+            The directory under which the manifest files resides.
+        """
+        self._dataset_name = dataset_name
+        self._manifest_dir = manifest_dir
+
+        self._splits = set()
+
+        for tsv_file in manifest_dir.glob("*.tsv"):
```
Comment (orena1) [inline src/fairseq2/datasets/speech.py:124]: This thing force to use all tsv in the folder, it there an option to load only few of them, either for debugging or for specific runs

File: src/fairseq2/recipes/wav2vec2/train.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+    """Load a :class:`Trainer` for wav2vec 2.0 model training."""
+    wall_watch = Stopwatch(start=True)
+
+    gang = setup_root_gang(log, monitored=config.monitored_gang)
+
+    checkpoint_manager = FileCheckpointManager(output_dir.joinpath("checkpoints"), gang)
+
+    # Load the dataset.
+    dataset_card = retrieve_asset_card(config.dataset)
+
```
Comment (orena1) [inline src/fairseq2/recipes/wav2vec2/train.py:188]: Do you need the extra empty line here?

File: src/fairseq2/recipes/wav2vec2/train.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    log.info("Loading {} speech dataset.", dataset_card.name)
+
+    dataset = load_speech_dataset(config.dataset)
+
```
Comment (orena1) [inline src/fairseq2/recipes/wav2vec2/train.py:192]: and also here? I think the logging can be line-attached to the function that it suprvise

File: src/fairseq2/recipes/wav2vec2/train.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        ddp_find_unused_parameters=False,
+        fsdp_skip_init=True,
+        fsdp_broadcast_state=not has_checkpoint,
+        fsdp_mixed_precision_dtype=config.dtype,
+        fsdp_fp32_reduce=True,
+        fsdp_wrap_granularity=config.fsdp_wrap_granularity,
+    )
+
+    if config.torch_compile:
+        model.encoder = compile_model(model.encoder, log)  # type: ignore[assignment]
```
Comment (orena1) [inline src/fairseq2/recipes/wav2vec2/train.py:235]: Why compile only the encoder?

**Comment by cbalioglu** [comment]
> Hi, thanks! General question about the dataloader: This assumes that all files are saved as individual audio files in the file system. Is this approach scalable for future use, especially as pre-training data requirements grow? Are there any plans to support other storage solutions, such as tar or parquet?

Hi @orena1, you are totally right. This PR is mainly intended for the parity work with the original fairseq implementation. For better scalability, we internally use two different techniques which we will eventually upstream to this recipe as well. The first technique uses non-compressed zip files to bundle multiple audio files together. We already have support for reading such files using `FileMapper` (see [here](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/data/data_pipeline.py#L471)). The other (more recent) technique is to leverage Parquet files instead of plain TSV files. We also have preliminary support for reading Parquet files in fairseq2 which we plan to extend in the near future. Unfortunately our documentation right now is almost non-existent, but starting next week we plan to invest a lot more time writing proper documentation.

File: src/fairseq2/datasets/speech.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    @abstractmethod
+    def splits(self) -> Set[str]:
+        """Return the set of splits."""
+
+
+load_speech_dataset = DelegatingDatasetLoader[SpeechDataset]()
+
+
+# TODO: FIX, INFER
+npc = 10
```
Comment (shagunsodhani) [inline src/fairseq2/datasets/speech.py:100]: what is this ?

File: src/fairseq2/recipes/wav2vec2/train.py (hunk 7 above)
Comment (shagunsodhani) [inline src/fairseq2/recipes/wav2vec2/train.py:235]: should this be `ddp_model.encoder` ?

File: src/fairseq2/datasets/speech.py (hunk 8 above)
Comment (kauterry) [reply src/fairseq2/datasets/speech.py:100]: Number of parallel calls for the collater.

File: src/fairseq2/recipes/wav2vec2/train.py (hunk 7 above)
Comment (kauterry) [reply src/fairseq2/recipes/wav2vec2/train.py:235]: @cbalioglu we should fix this in the ASR pipeline too right?

**Comment by orena1** [comment]
> > Hi, thanks! General question about the dataloader: This assumes that all files are saved as individual audio files in the file system. Is this approach scalable for future use, especially as pre-training data requirements grow? Are there any plans to support other storage solutions, such as tar or parquet?
> 
> Hi @orena1, you are totally right. This PR is mainly intended for the parity work with the original fairseq implementation. For better scalability, we internally use two different techniques which we will eventually upstream to this recipe as well. The first technique uses non-compressed zip files to bundle multiple audio files together. We already have support for reading such files using `FileMapper` (see [here](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/data/data_pipeline.py#L471)). The other (more recent) technique is to leverage Parquet files instead of plain TSV files. We also have preliminary support for reading Parquet files in fairseq2 which we plan to extend in the near future. Unfortunately our documentation right now is almost non-existent, but starting next week we plan to invest a lot more time writing proper documentation.

Thanks a lot! happy to see how it progress, and how exactly you handle random access with zip, thanks [🙏](https://emojipedia.org/folded-hands)

File: src/fairseq2/models/wav2vec2/frontend.py [hunk 9]
```diff
@@ -161,9 +161,11 @@ def extract_features(
         if self.feature_extractor is not None:
             seqs, padding_mask = self.feature_extractor(seqs, padding_mask)
 
+        feature_penalty = seqs.float().pow(2).mean()
```
Comment (cbalioglu) [inline src/fairseq2/models/wav2vec2/frontend.py:164]: I suggest:

```python
raw_features = seqs

seqs = self.post_extract_layer_norm(seqs)

return seqs, padding_mask, raw_features
```

and computing the `float().pow(2).mean()` (or better yet use `torch.linalg.vector_norm()`) only during loss calculation if the penalty weight is greater than zero.

File: src/fairseq2/models/wav2vec2/frontend.py [hunk 10]
```diff
@@ -129,15 +129,15 @@ def forward(
...
+        seqs, padding_mask, _ = self.extract_features(seqs, padding_mask)
 
         seqs, padding_mask, _ = self.process_features(seqs, padding_mask)
 
         return seqs, padding_mask
 
     def extract_features(
         self, seqs: Tensor, padding_mask: Optional[PaddingMask]
-    ) -> Tuple[Tensor, Optional[PaddingMask]]:
+    ) -> Tuple[Tensor, Optional[PaddingMask], Tensor]:
```
Comment (cbalioglu) [inline src/fairseq2/models/wav2vec2/frontend.py:140]: We should coordinate with Arrival since this would be a breaking change. @shagunsodhani

File: src/fairseq2/models/wav2vec2/model.py [hunk 11]
```diff
@@ -42,6 +42,8 @@ class Wav2Vec2Model(Model):
     final_target_proj: Linear
     num_distractors: int
     logit_temp: float
+    diversity_loss_weight: float
```
Comment (cbalioglu) [inline src/fairseq2/models/wav2vec2/model.py:45]: See my comment below. I think loss weights should only be specified as parameters of `compute_loss()`.

File: src/fairseq2/models/wav2vec2/model.py [hunk 12]
```diff
@@ -124,19 +130,23 @@ def extract_features(self, batch: SequenceBatch) -> Wav2Vec2Features:
         :param batch:
             The batch of sequences to process.
         """
-        seqs, padding_mask, targets, temporal_mask = self.run_frontend(
+        seqs, padding_mask, targets, temporal_mask, feature_penalty = self.run_frontend(
```
Comment (cbalioglu) [inline src/fairseq2/models/wav2vec2/model.py:133]: Again, this has to be coordinated with Arrival. @shagunsodhani

File: src/fairseq2/metrics/accuracy.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+import torch
+from torch import Tensor
+from torcheval.metrics import Metric
+from typing_extensions import Self
+
+from fairseq2.typing import Device, override
+
+
+@final
+class AccuracyMetric(Metric[Tensor]):
```
Comment (cbalioglu) [inline src/fairseq2/metrics/accuracy.py:20]: Any chance of using one of the implementations in https://github.com/pytorch/torcheval/blob/main/torcheval/metrics/classification/accuracy.py instead of this implementation?

File: src/fairseq2/models/wav2vec2/frontend.py (hunk 9 above)
Comment (shagunsodhani) [reply src/fairseq2/models/wav2vec2/frontend.py:164]: +1 to returning `raw_features` (and not `feature_penalty`). @cbalioglu do you want to consider returning an object here (instead of a tuple of tensors) ?

File: src/fairseq2/models/wav2vec2/model.py [hunk 14]
```diff
@@ -359,7 +376,13 @@ class Wav2Vec2Output:
...
     sequence length."""
 
-    def compute_loss(self, *, diversity_loss_weight: float = 0.1) -> Wav2Vec2Loss:
+    feature_penalty: Tensor
+
+    diversity_loss_weight: float = 0.1
+
+    feature_penalty_weight: float = 10.0
+
+    def compute_loss(self) -> Wav2Vec2Loss:
```
[the remark this answers is not in the document]
Comment (kauterry) [reply src/fairseq2/models/wav2vec2/model.py:385]: These parameters change with different configurations of wav2vec2, for instance the large_lv60k wav2vec2 pretraining config sets the feature_penalty_weight to 0.0. How do we achieve this?

File: src/fairseq2/models/wav2vec2/frontend.py (hunk 9 above)
Comment (cbalioglu) [reply src/fairseq2/models/wav2vec2/frontend.py:164]: Yes, I think it makes sense to convert this to a proper dataclass instead of a tuple at this point. Once we land these changes, let's make sure that we reflect them in your repo. We are pretty close to v0.3 release, it would be good timing.

File: src/fairseq2/models/wav2vec2/model.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (cbalioglu) [reply src/fairseq2/models/wav2vec2/model.py:385]: If we include them in recipe configuration, we can have different values for different presets.

File: src/fairseq2/nn/utils/mask.py [hunk 15]
```diff
@@ -171,3 +171,20 @@ def _generate_mask(indices: Tensor, max_row_len: int) -> Tensor:
     bool_mask = torch.full_like(float_mask, False, dtype=torch.bool)
 
     return bool_mask.scatter_(1, indices, True)
+
+
+def get_num_masks(mask: Tensor) -> Tensor:
```
Comment (kauterry) [inline src/fairseq2/nn/utils/mask.py:176]: This is a convenience function which returns the number of masks present in a mask tensor of shape (bsz, seq_len). This is implemented in an efficient way, I believe this might be useful for some users.

File: src/fairseq2/metrics/accuracy.py (hunk 13 above)
Comment (kauterry) [reply src/fairseq2/metrics/accuracy.py]: Done! Thank you for the pointer, removed this unnecessary Metric.

**Comment by cbalioglu** [comment]
Let's keep the branch around for further data loading experiments, but closing this PR. Please continue doing further development on team repo.
